### PR TITLE
ICU-21357 Support IEC Binary Prefixes in MeasureUnit

### PR DESCRIPTION
**Pull request by hugovdm**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21357
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [x] Documentation is changed or added

Resolves: https://github.com/icu-units/icu/issues/114

### CODE REVIEW

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [inline icu4c/source/i18n/unicode/measunit.h:88]: The "+ 0" here intended to remind that this is 10^0.

Comment (younies) [inline icu4c/source/i18n/unicode/measunit.h:88]: Instead of using "30" as an offset for the SI prefixes. I prefer using a constant called for example `SI_PREFIX_OFFSET` which equals 30 or whatever value.

Then use this constant as an offset for all the other SI prefixes.

File: icu4c/source/i18n/unicode/measunit.h
Comment (younies) [inline icu4c/source/i18n/unicode/measunit.h:95]: as suggested use `SI_PREFIX_OFFSET` instead of `UMEASURE_PREFIX_ONE`

File: icu4c/source/i18n/unicode/measunit.h
Comment (younies) [inline icu4c/source/i18n/unicode/measunit.h:109]: ditto

and for all the other SI prefixes

File: icu4c/source/i18n/unicode/measunit.h
Comment (younies) [inline icu4c/source/i18n/unicode/measunit.h:249]: Rename to be `BINARY_PREFIX_OFFSET`

OR any other name that means it is a constant and an offset.

File: icu4c/source/test/intltest/units_test.cpp
Comment (younies) [inline icu4c/source/test/intltest/units_test.cpp:202]: +1

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (younies) [inline icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:132]: I prefer instead of using the number. use a constant. 
Using constant is more illustrative and error-prone. it is easy to miss type a number.

for example: `MeasurePrefix.SI_BASE`

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (younies) [inline icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:140]: ditto 

and all other SI prefixes

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (younies) [inline icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:300]: I prefer instead of using the number. use a constant. 
Using constant is more illustrative and error-prone. it is easy to miss type a number.

for example: `MeasurePrefix.BINARY_BASE`

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java
Comment (younies) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:3531]: +1

**Review by younies — DISMISSED** [review]

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [reply icu4c/source/i18n/unicode/measunit.h:249]: This was the name that was suggested in the meeting, and approved. I believe I must not change it without another ICU-TC review.

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [reply icu4c/source/i18n/unicode/measunit.h:88]: I like limiting how much cruft is present in a public header file. Another constant means another exported symbol accessible to clients. This was discussed in the ICU-TC meeting, so reviewed and approved by the committee. I believe further changes  (to the public header file) would need another review.

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:132]: This is also a public class - though, Java at least has private members. I hear what you're saying, but again: committee review. (At least I think it's harder to "mistype" numbers like 10 and 1024, it's tested in the unit tests, and client code doesn't need to be typing these numbers...)

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [reply icu4c/source/i18n/unicode/measunit.h:95]: UMEASURE_PREFIX_ is the symbol prefix that indicates this is an enum value of the UMeasurePrefix type. This is the style of the public ICU4C APIs - partially also because C does not have namespaces, so public symbols need to be very clear. Again: this is the public API that was approved by the committee.

File: icu4c/source/i18n/measunit_extra.cpp
Comment (sffc) [inline icu4c/source/i18n/measunit_extra.cpp:653]: Please include the C API boilerplate.

```suggestion
U_CAPI int32_t U_EXPORT2
umeas_getPrefixPower(UMeasurePrefix unitPrefix) {
```

File: icu4c/source/i18n/measunit_extra.cpp
Comment (sffc) [inline icu4c/source/i18n/measunit_extra.cpp:663]:
```suggestion
U_CAPI int32_t U_EXPORT2
umeas_getPrefixBase(UMeasurePrefix unitPrefix) {
```

File: icu4c/source/i18n/unicode/measunit.h
Comment (sffc) [inline icu4c/source/i18n/unicode/measunit.h:75]: You have a `-` after kilo and nano but not mebi

File: icu4c/source/i18n/unicode/measunit.h
Comment (sffc) [reply icu4c/source/i18n/unicode/measunit.h:88]: For public use we want people to use the new C functions umeas_getPrefixPower and umeas_getPrefixBase.  I don't feel strongly about whether to introduce `UPRV_SI_PREFIX_OFFSET` as an internal thing.  If we did, we could easily label it `@internal` to avoid an additional API review.

File: icu4c/source/i18n/unicode/measunit.h
Comment (sffc) [reply icu4c/source/i18n/unicode/measunit.h:95]: See above for comment about `UPRV_SI_PREFIX_OFFSET`

File: icu4c/source/i18n/unicode/measunit.h
Comment (sffc) [reply icu4c/source/i18n/unicode/measunit.h:249]: See above for comment about `UPRV_SI_PREFIX_OFFSET`

File: icu4c/source/i18n/units_converter.cpp
Comment (sffc) [inline icu4c/source/i18n/units_converter.cpp:312]: kilokelvin is fine; it could be used to describe temperatures at the surface of the sun, for example.  I don't think I've ever seen a kilocelsius or even have a guess about what that could mean.

We should, at least, have a test case that demonstrates what our behavior is in edge cases like this.  Follow-up issue?

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:132]: I'm in favor of just writing 10 here on the Java side.  This code will never need to be changed, and we are doing a review right now to make sure it is correct.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java
Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java:221]: Why capital-I `Integer` rather than `int`?

Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java:221]: Why capital-I `Integer` rather than `int`?

**Review by sffc — COMMENTED** [review]
LGTM from me with some optional small comments

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:132]: thanks. I'm leaving this as just "10" and "1024", trusting conversion unit tests, prefix unit tests, and code review to ensure they're right.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java:221]: Heh, "I have no idea": I can't remember what inspired this.

(I find one for-loop using "Integer" in this file - should I change that to "int" too? I don't think that one was my inspiration though!)

Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java:221]: Dupe.

File: icu4c/source/i18n/units_converter.cpp
Comment (hugovdm) [reply icu4c/source/i18n/units_converter.cpp:312]: I added a ticket for tracking handling of "undefined behaviour" - https://github.com/icu-units/icu/issues/127 - additional items can be added to it in the future.

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [reply icu4c/source/i18n/unicode/measunit.h:88]: For now, I'm leaving this as just UMEASURE_PREFIX_ONE - there's only one copy of "30", so it's really just a question of whether this constant needs an alias. An alias would make it clearer where the arbitrary value is used (here and umeas_getPrefixPower), but I think the number of references to UMEASURE_PREFIX_ONE is not too many.

I did look at how vscode handles documentation for the symbol, and decided thereby that the "arbitrary value" comment should rather be combined with the main documentation for this symbol: I went adhead and did this.

File: icu4c/source/i18n/unicode/measunit.h
Comment (hugovdm) [reply icu4c/source/i18n/unicode/measunit.h:75]: Done.

File: icu4c/source/i18n/measunit_extra.cpp
Comment (hugovdm) [reply icu4c/source/i18n/measunit_extra.cpp]: Done though I do suspect the C APIs won't be very useful, given that there's no C API that's actually making use of this enum? Leaving unresolved...

**Review by sffc — DISMISSED** [review]
LGTM please squash (and check if younies has any more comments)

**Review by younies — APPROVED** [review]

**Comment by hugovdm** [comment]
FYI: github-suggestions are squashbot safe, because the author is still the PR author. Example: fbb1ad2a96523a78522ce86811a25d0325474166
Shows also how co-authors can be indicated:
Co-authored-by: Shane F. Carr <shane@unicode.org>

I'll drop a note of this on the squashbot issue tracker.

**Review by younies — APPROVED** [review]

**Pull request merged**
